File: sdk/typescript/src/namespaces/dag-operations.ts
/**
 * DAG Operations Namespace API
 *
 * Provides operations for pipeline DAG node automation and graph-level flows.
 */

import type { AragoraClient } from '../client';

export interface DagOperationResult {
  success: boolean;
  message: string;
  created_nodes: string[];
  metadata: Record<string, unknown>;
}

export interface DagGraphResponse {
  data: Record<string, unknown>;
}

export interface DagOperationResponse {
  data: DagOperationResult;
}

export interface DebateNodeOptions {
  agents?: string[];
  rounds?: number;
}

export interface AssignAgentsOptions {
  node_ids?: string[];
}

export interface FindPrecedentsOptions {
  max_results?: number;
}

export interface ClusterIdeasOptions {
  threshold?: number;
}

export interface AutoFlowOptions {
  config?: Record<string, unknown>;
}

export class DagOperationsNamespace {
  constructor(private client: AragoraClient) {}

  async getGraph(graphId: string): Promise<DagGraphResponse> {
    return this.client.request<DagGraphResponse>(
      'GET',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}`
    );
  }

  async debateNode(
    graphId: string,
    nodeId: string,
    options?: DebateNodeOptions,
  ): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/nodes/${encodeURIComponent(nodeId)}/debate`,
      {
        body: {
          ...(options?.agents ? { agents: options.agents } : {}),
          rounds: options?.rounds ?? 3,
        },
      },
    );
  }

  async decomposeNode(graphId: string, nodeId: string): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/nodes/${encodeURIComponent(nodeId)}/decompose`,
    );
  }

  async prioritizeNode(graphId: string, nodeId: string): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/nodes/${encodeURIComponent(nodeId)}/prioritize`,
    );
  }

  async assignAgents(
    graphId: string,
    nodeId: string,
    options?: AssignAgentsOptions,
  ): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/nodes/${encodeURIComponent(nodeId)}/assign-agents`,
      {
        body: options?.node_ids ? { node_ids: options.node_ids } : {},
      },
    );
  }

  async executeNode(graphId: string, nodeId: string): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/nodes/${encodeURIComponent(nodeId)}/execute`,
    );
  }

  async findPrecedents(
    graphId: string,
    nodeId: string,
    options?: FindPrecedentsOptions,
  ): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/nodes/${encodeURIComponent(nodeId)}/find-precedents`,
      {
        body: { max_results: options?.max_results ?? 5 },
      },
    );
  }

  async clusterIdeas(
    graphId: string,
    ideas: string[],
    options?: ClusterIdeasOptions,
  ): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/cluster-ideas`,
      {
        body: {
          ideas,
          threshold: options?.threshold ?? 0.3,
        },
      },
    );
  }

  async autoFlow(
    graphId: string,
    ideas: string[],
    options?: AutoFlowOptions,
  ): Promise<DagOperationResponse> {
    return this.client.request<DagOperationResponse>(
      'POST',
      `/api/v1/pipeline/dag/${encodeURIComponent(graphId)}/auto-flow`,
      {
        body: {
          ideas,
          ...(options?.config ? { config: options.config } : {}),
        },
      },
    );
  }
}

File: sdk/python/aragora_sdk/namespaces/dag_operations.py
"""DAG operations namespace API."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class DAGOperationsAPI:
    """Synchronous DAG operations API."""

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_graph(self, graph_id: str) -> dict[str, Any]:
        """Get a DAG graph by ID."""
        return self._client.request("GET", f"/api/v1/pipeline/dag/{graph_id}")

    def debate_node(
        self,
        graph_id: str,
        node_id: str,
        *,
        agents: list[str] | None = None,
        rounds: int = 3,
    ) -> dict[str, Any]:
        """Run a debate operation for a node."""
        payload: dict[str, Any] = {"rounds": rounds}
        if agents is not None:
            payload["agents"] = agents
        return self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/debate",
            json=payload,
        )

    def decompose_node(self, graph_id: str, node_id: str) -> dict[str, Any]:
        """Decompose a node into child nodes."""
        return self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/decompose",
        )

    def prioritize_node(self, graph_id: str, node_id: str) -> dict[str, Any]:
        """Prioritize children for a node."""
        return self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/prioritize",
        )

    def assign_agents(
        self,
        graph_id: str,
        node_id: str,
        *,
        node_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        """Assign agents for a node or a list of node IDs."""
        payload: dict[str, Any] = {}
        if node_ids is not None:
            payload["node_ids"] = node_ids
        if payload:
            return self._client.request(
                "POST",
                f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/assign-agents",
                json=payload,
            )
        return self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/assign-agents",
        )

    def execute_node(self, graph_id: str, node_id: str) -> dict[str, Any]:
        """Execute a node."""
        return self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/execute",
        )

    def find_precedents(
        self,
        graph_id: str,
        node_id: str,
        *,
        max_results: int = 5,
    ) -> dict[str, Any]:
        """Find precedent nodes for a node."""
        return self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/find-precedents",
            json={"max_results": max_results},
        )

    def cluster_ideas(
        self,
        graph_id: str,
        ideas: list[str],
        *,
        threshold: float = 0.3,
    ) -> dict[str, Any]:
        """Cluster a set of ideas into graph nodes."""
        return self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/cluster-ideas",
            json={"ideas": ideas, "threshold": threshold},
        )

    def auto_flow(
        self,
        graph_id: str,
        ideas: list[str],
        *,
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Run the end-to-end DAG auto-flow operation."""
        payload: dict[str, Any] = {"ideas": ideas}
        if config is not None:
            payload["config"] = config
        return self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/auto-flow",
            json=payload,
        )


class AsyncDAGOperationsAPI:
    """Asynchronous DAG operations API."""

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_graph(self, graph_id: str) -> dict[str, Any]:
        """Get a DAG graph by ID."""
        return await self._client.request("GET", f"/api/v1/pipeline/dag/{graph_id}")

    async def debate_node(
        self,
        graph_id: str,
        node_id: str,
        *,
        agents: list[str] | None = None,
        rounds: int = 3,
    ) -> dict[str, Any]:
        """Run a debate operation for a node."""
        payload: dict[str, Any] = {"rounds": rounds}
        if agents is not None:
            payload["agents"] = agents
        return await self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/debate",
            json=payload,
        )

    async def decompose_node(self, graph_id: str, node_id: str) -> dict[str, Any]:
        """Decompose a node into child nodes."""
        return await self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/decompose",
        )

    async def prioritize_node(self, graph_id: str, node_id: str) -> dict[str, Any]:
        """Prioritize children for a node."""
        return await self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/prioritize",
        )

    async def assign_agents(
        self,
        graph_id: str,
        node_id: str,
        *,
        node_ids: list[str] | None = None,
    ) -> dict[str, Any]:
        """Assign agents for a node or a list of node IDs."""
        payload: dict[str, Any] = {}
        if node_ids is not None:
            payload["node_ids"] = node_ids
        if payload:
            return await self._client.request(
                "POST",
                f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/assign-agents",
                json=payload,
            )
        return await self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/assign-agents",
        )

    async def execute_node(self, graph_id: str, node_id: str) -> dict[str, Any]:
        """Execute a node."""
        return await self._client.request(
            "POST", f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/execute",
        )

    async def find_precedents(
        self,
        graph_id: str,
        node_id: str,
        *,
        max_results: int = 5,
    ) -> dict[str, Any]:
        """Find precedent nodes for a node."""
        return await self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/nodes/{node_id}/find-precedents",
            json={"max_results": max_results},
        )

    async def cluster_ideas(
        self,
        graph_id: str,
        ideas: list[str],
        *,
        threshold: float = 0.3,
    ) -> dict[str, Any]:
        """Cluster a set of ideas into graph nodes."""
        return await self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/cluster-ideas",
            json={"ideas": ideas, "threshold": threshold},
        )

    async def auto_flow(
        self,
        graph_id: str,
        ideas: list[str],
        *,
        config: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Run the end-to-end DAG auto-flow operation."""
        payload: dict[str, Any] = {"ideas": ideas}
        if config is not None:
            payload["config"] = config
        return await self._client.request(
            "POST",
            f"/api/v1/pipeline/dag/{graph_id}/auto-flow",
            json=payload,
        )
